bitcoin/src/address/error.rs

```diff
@@ -31,14 +80,72 @@ pub enum Error {
     WitnessVersion(witness_version::Error),
     /// A witness program error.
     WitnessProgram(witness_program::Error),
+}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParseError::*;
+
+        match *self {
+            Base58(ref e) => write_err!(f, "base58 address encoding error"; e),
+            Bech32(ref e) => write_err!(f, "bech32 address encoding error"; e),
+            EmptyBech32Payload => write!(f, "the bech32 payload was empty"),
+            InvalidBech32Variant { expected, found } => write!(
+                f,
+                "invalid bech32 checksum variant found {:?} when {:?} was expected",
+                found, expected
+            ),
+            WitnessVersion(ref e) => write_err!(f, "witness version conversion/parsing error"; e),
+            WitnessProgram(ref e) => write_err!(f, "witness program error"; e),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ParseError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use ParseError::*;
+
+        match *self {
+            Base58(ref e) => Some(e),
+            Bech32(ref e) => Some(e),
+            WitnessVersion(ref e) => Some(e),
+            WitnessProgram(ref e) => Some(e),
+            EmptyBech32Payload | InvalidBech32Variant { .. } => None,
+        }
+    }
+}
+
+impl From<base58::Error> for ParseError {
+    fn from(e: base58::Error) -> Self { Self::Base58(e) }
+}
```

Comment:
If that is the case then I am misunderstanding the point of the `Error` type. As I understood it the type was for users of the lib to use as a return value for functions they write that call multiple functions from the address module/type.  There is no use of it in the codebase on this branch. Or am I missing something?

Reply:
I agree with @apoelstra. For some reason I thought that `.into()` would do a transitive call from base58::Error -> ParseError -> Error, but it does not. (and for good reasons!)

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4ee5670908a74cd6fdf709799f8cf2be

Reply:
I am reconsidering this approach as a whole now. It is not true that `base58::Error` must go via `address::Error`. Consider the following code where this leads to surprising behaviour because of our from impl. 

```rust
pub fn user_fn() -> Result<(), address::Error> {
    // This will convert to address::Error variant, but this has nothing to do with address. 
	let base58_parsed = bytes::from_base58()?;
}
```

I guess we should only do it for obvious cases where we are clearly defined error heararchy. (which is not the case with base58).

Reply:
> As I understood it the type was for users of the lib to use as a return value for functions they write that call multiple functions from the address module/type.

Right, but say they call multiple functions from the module *and also* call a function that returns a `base58::Error`. Now they need to write `??` (which won't compile) or `.map_err(Error::from)?` (which is ugly).

Reply:
>I guess we should only do it for obvious cases where we are clearly defined error heararchy. (which is not the case with base58).

Ah, yeah, I do agree with this. I had the same thought/fear when working this out in my head.

Reply:
> Right, but say they call multiple functions from the module and also call a function that returns a `base58::Error`.

I would think in this case the user should be defining a new error type or just using `anyhow::Result`. Actually now I write this only libs create error types anyway and library writers using `rust-bitcoin` will be used to creating error types, they do not need this to be in `rust-bitcoin`. Application devs just use `anyhow` so they don't need it either.